### DISCUSSION TITLE
How to use esp32 to decoder BMS Jikong BD6A20S8P

### DISCUSSION THREAD

**Discussion by NguyenHau101202** [syssi/esphome-jk-bms#500] [General]
hello everyone, i';m a newbie, i have a project about use ESp32 read the data of BMS Jikong through RS485 protocol. but i don't have any reference document about frame data structure of it. Can someone share it with me ? Thanks

**Comment by meccip** [comment]
i think all info about protocol is in the docs folder in code source https://github.com/syssi/esphome-jk-bms/tree/main/docs

**Comment by NguyenHau101202** [comment]
i saw 2 file : RS485 English Obsolete and RS485 communication example. can someone tell me the different of both file. because the data frame structure of 2 file is different. i don't know which one to use

**Reply by meccip** [reply]
but what do you want to do ?

**Reply by syssi** [reply]
I depends. ;-) If you set the `UART protocol` to `4G-GPS Remote` at the JK-BMS app this YAML should be able to talk to the BMS: https://github.com/syssi/esphome-jk-bms/blob/main/esp32-example.yaml

If you set the `UART protocol` to `JK BMS RS485 Modbus` this YAML / protocol description should apply: https://github.com/syssi/esphome-jk-bms/pull/409/files

You could also communicate using bluetooth with your BMS (this requires a ESP32): https://github.com/syssi/esphome-jk-bms/blob/main/esp32-ble-example.yaml

![324207556-fada8f59-0599-4341-af51-68909da503ac](https://github.com/syssi/esphome-jk-bms/assets/2735933/1bd25c71-3ad7-4358-bff1-1a93588a48b0)

**Reply by meccip** [reply]
hi @syssi  can you help me with some info ? i want to get max/min cell number from bms with mod bus https://github.com/syssi/esphome-jk-bms/pull/409/files#diff-af9279656260549e78da3780c089c1e1adc3a34a19220f0e86e3f7a6c32412e6 
There are at the same address. Can you help me ?

 0x1248   72    UINT8/UINT8 2  R    MaxVolCellNbr / MinVolCellNbr
  - platform: modbus_controller
    modbus_controller_id: bms0
    name: "${name} max voltage cell"
    address: 0x1248
    register_type: holding
    value_type: U_WORD               ??????
    unit_of_measurement: "V"
    device_class: voltage
    state_class: measurement
    accuracy_decimals: 0

**Reply by syssi** [reply]
@meccip You have to extract the hi/lo byte using a bitmask. Could you create another discussion to address this task and describe your setup? Thanks in advance.

**Comment by NguyenHau101202** [comment]
i use esp32 to read the data of bms and export it TFT display. i wanna decode the data rs485

**Comment by NguyenHau101202** [comment]
Can someone tell me the difference of 2 file : RS485 English Obsolete and RS485 communication example ? please!

**Reply by syssi** [reply]
The [RS485-english-OBSOLETE.pdf](https://github.com/syssi/esphome-jk-bms/blob/main/docs/RS485-english-OBSOLETE.pdf) is a protocol description of a firmware version / BMS models from many years ago. This protocol doesn't apply anymore and can be ignored.

The [RS485 Communication example.pdf](https://github.com/syssi/esphome-jk-bms/blob/main/docs/RS485%20Communication%20example.pdf) is an example request/response (cell info frame) of the `4G-GPS Remote` protocol.

**Reply by syssi** [reply]
You should focus on this protocol description: https://github.com/syssi/esphome-jk-bms/blob/add-jk-pbx-support/docs/pb2a16s20p/JK_BMS.RS485.Modbus.v1_0.pdf

A first draft of the implementation is available here: https://github.com/syssi/esphome-jk-bms/blob/add-jk-pbx-support/esp32-jk-pb-modbus-example.yaml

**Reply by NguyenHau101202** [reply]
may i ask whether BMS RS485 Modbus v1 document also use for BMS jikong bd6a20s8p ?

**Comment by NguyenHau101202** [comment]
Is this file also use for jk BD6A20S8P ?

**Reply by syssi** [reply]
Which firmware/hardware version do you own?

**Reply by NguyenHau101202** [reply]
how can i find those information ? i'm a newbie. i just bought a BMS Jikong and i I don't understand it clearly

**Reply by syssi** [reply]
Please download the Android App as first step! The is a "about view" which mentions some details about the BMS. I guess your BMS supports BLE communication and the "GPS UART-TTL protocol" only.

**Reply by NguyenHau101202** [reply]
my bms has rs485 protocol, and i have a project about decoding it. but there are many document about dataframe structure . i don't know which one to use

**Reply by syssi** [reply]
Which connector do you use to talk to your BMS?

**Reply by NguyenHau101202** [reply]
i use rs485 protocol, connect to a RS485 TTL to UART and connect the RS485 TTL module with an esp32.

**Reply by syssi** [reply]
I assume we are talking about this connector:

![rs485](https://github.com/syssi/esphome-jk-bms/assets/2735933/8773de00-a6a9-4609-bd7f-9457736a799f)

Please install the Android app, check the firmware/hardware version and check the settings page. What's available here?

**Reply by meccip** [reply]
i think is easy to buy a lcd 4.3 inch for bms
![20231230_003557](https://github.com/syssi/esphome-jk-bms/assets/10565481/7e7ba45a-4f0c-451f-a80a-421191aaa74f)

**Reply by NguyenHau101202** [reply]
i have a project at my university about building a DIY decoder for it

**Reply by NguyenHau101202** [reply]
![438083845_1676918829712384_3504229600917342600_n](https://github.com/syssi/esphome-jk-bms/assets/163629747/79ffec47-fc99-44bb-90de-58c298cd03e3)
![438154342_2170548329960184_7863523384262615178_n](https://github.com/syssi/esphome-jk-bms/assets/163629747/14f37461-6080-4c96-8743-3172ebbf642f)

this is my bms and i don't know data frame structure of it

**Reply by syssi** [reply]
Please install the Android app and provide the firmware/hardware version as next step!

**Reply by meccip** [reply]
If you want to powerup bms. Install red wires and connect black wire (B-) to GND and B+(last red wire) to +19v minimum. Also put GND to B-(from 6mm Screw). Push the button to start bms.

**Reply by NguyenHau101202** [reply]
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/66ea0927-b622-421e-a072-12acffab7fa0)
I use this model, how can i find frame data structure of this protocol model ? Rs485 or can is ok

**Reply by syssi** [reply]
Could you provide a screenshot of all views?

**Reply by NguyenHau101202** [reply]
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/eba299b7-c8eb-4888-a224-9db12f9115c9)
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/b36ffef5-c847-4be3-b1bc-42e3d3613963)
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/da2404b6-11a2-4bc7-b998-b606942a88ea)
This my bms app. So, if i wanna use esp32 to communicate with my bms through rs485 protocol. Which document should i use( about data frame, register map)?

**Reply by syssi** [reply]
The most important screenshot (control view) is missing and please check the advanced settings of the settings view. Can you choose between different protocols here?

**Reply by NguyenHau101202** [reply]
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/3cffd43b-1cf8-4fcb-9cec-7a11cdbf7ac2)
What should i do next?

**Reply by syssi** [reply]
If you open the `UART1 Protocol No` drop-down which values/protocols are available here?

**Reply by NguyenHau101202** [reply]
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/c5c3c5b9-e2a8-421f-9875-4297ca13da67)
It's here

**Reply by NguyenHau101202** [reply]
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/a35ee2e2-c5af-4eb9-ad25-2dbe1e536529)
And i have a question. I saw in the document, P4 is RS485 and P5 is GPS, whether i should use p4 or p5 with a rs485 module. Because i don't have the connector cable of p4. Thanks

**Reply by syssi** [reply]
It looks like your BMS supports the `4G GPS Remote module communcation` protocol:

https://github.com/syssi/esphome-jk-bms/blob/main/docs/RS485%20Communication%20example.pdf
https://github.com/syssi/esphome-jk-bms/blob/main/docs/protocol-design.md
https://github.com/syssi/esphome-jk-bms/blob/main/esp8266-example.yaml

And the `JK BMS RS485 Modbus V1.0` protocol:

https://github.com/syssi/esphome-jk-bms/blob/main/docs/pb2a16s20p/JK_BMS.RS485.Modbus.v1_0.pdf
https://github.com/syssi/esphome-jk-bms/blob/main/esp32-jk-pb-modbus-example.yaml

on port `P5`. IMO `P5` is the UART1 port. I don't know much about `P4`. This is probably the setting `CAN protocol no`. Which protocols are available here?

**Reply by NguyenHau101202** [reply]
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/39428b21-ce77-4aec-912c-308562953f6b)
It's here

**Reply by NguyenHau101202** [reply]
And now if i wanna use p5 for rs485 protocol. I must have this one, right ? ![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/db688234-9e63-4121-b3b2-6573ad4a3adf)

**Reply by syssi** [reply]
Yes, you have to use P5. If you use the dongle the UART-TTL signal of P5 is transformed into an RS485 signal. You have to attach another RS485 converter to your ESP to transform the signal back to UART-TTL. If you don't need the optical isolation and if you don't power the ESP using your battery back you could skip the RS485 layer and attach the BMS directly to the ESP.

**Reply by NguyenHau101202** [reply]
I don't understand this on "If you don't need the optical isolation and if you don't power the ESP using your battery back you could skip the RS485 layer and attach the BMS directly to the ESP" . What do you mean ?

**Reply by meccip** [reply]
@NguyenHau101202  did you have the rs485 adapter from the picture already ?

**Reply by NguyenHau101202** [reply]
Yes , i have. What should i do next ? Thanks

**Reply by meccip** [reply]
![20240510_162844](https://github.com/syssi/esphome-jk-bms/assets/10565481/ad9acfd1-ece1-4fae-8e6c-2ef45942d3f2)
Your adapter does fit in rs485 connector on bms ?

**Reply by NguyenHau101202** [reply]
You mean fit about shape of connector ? Or another thing

**Reply by meccip** [reply]
adapter from the photo must fit into rs485 bms connector.
But if you use that adaptor you must add a new RStoTTL device to connect to TTL. (What syssi try to tell you.)

So, i think it is much simple to conect esp directly to RS485 bms connector (because there is already a RxTx signal)

**Reply by meccip** [reply]
it is described at schematics area, in the first page of code https://github.com/syssi/esphome-jk-bms?tab=readme-ov-file#schematics

**Reply by meccip** [reply]
Very important !! Dont use v+ from bms connector. Use 5v external supply to power the esp, just wire GND with bms.

**Reply by NguyenHau101202** [reply]
As you said, if i connect directly my esp32 to bms . What document should i use to program ?

**Reply by syssi** [reply]
ESP8266 <-> UART-TTL GPS port (`P5`) + `4G-GPS Remote module communication` protocol enabled: https://github.com/syssi/esphome-jk-bms/blob/main/esp8266-example.yaml
ESP32 <-> UART-TTL GPS port (`P5`) + `4G-GPS Remote module communication` protocol enabled: https://github.com/syssi/esphome-jk-bms/blob/main/esp32-example.yaml

**Reply by NguyenHau101202** [reply]
I mean the document about protocol and frame structure, register map..

**Reply by syssi** [reply]
Did you read this comment? https://github.com/syssi/esphome-jk-bms/discussions/500#discussioncomment-9380581

**Reply by NguyenHau101202** [reply]
Yeah, but i wanna build DIY decoder, with TFT display. I build base on adruino platform iíntead of esphome.

**Reply by syssi** [reply]
Again: These are the protocol details you are looking for: https://github.com/syssi/esphome-jk-bms/blob/main/docs/protocol-design.md

**Reply by NguyenHau101202** [reply]
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/8cccc82a-c0bb-4053-892a-f3b3a5494a6d)
![image](https://github.com/syssi/esphome-jk-bms/assets/163629747/7788c132-9f5d-4988-ab97-bc4c1f4809f1)
I just build a program to read data of bms. I added a module rs485 ttl. But, i upload the code, the tx led in rs485 ttl is blink, the rx led is not. Nothing response. Help me

**Reply by syssi** [reply]
You could use this project to verify your wiring and to make sure your BMS is able to respond to valid requests.

**Reply by meccip** [reply]
i use the same rs ttl board. I can say that there is a possibility that the board is broken. I really experienced this, only one led worked. I changed the board and it worked. Or if the right protocol is not selected into the bms, the led will blink the same.
